### DISCUSSION TITLE
if its avaliable to excute lambda(cloudtrail) in multi aws accounts by using tool c7n-org?

### DISCUSSION THREAD

**Discussion by youzhoutai** [cloud-custodian/cloud-custodian#7317] [General]
I am trying to deploy lambda in multi accounts(B and C). Major account A, other account B and C. And 1.one config file already defined as shown below:
accounts:
- account_id: 'B'
  name: account-B
  regions:
  - cn-north-1
  role: arn:aws-cn:iam::B:role/CustodianEC2Tagging
  tags:
  - type:test
  - scope:B
- account_id: 'C'
  name: account-C
  regions:
  - cn-north-1
  role: arn:aws-cn:iam::C:role/CustodianEC2Tagging
  tags:
  - type:test
  - scope:C
2. the policy file as shown below:
policies:
  - name: ec2-tag-running
    resource: ec2
    mode:
      type: cloudtrail
      events:
       - source: ec2.amazonaws.com
         event: RunInstances
         ids: "responseElements.instancesSet.items[].instanceId"
    actions:
      - type: tag
        key: foo
        value: bar
3. The output after the excution

c7n-org run -c custodian-all-cnf.yml -s output -u ec2tagging.yml --cache-period 0
2018-03-05 09:23:10,543: c7n_org:ERROR Exception running policy:ec2-tag-running account:account-301745419360 region:cn-north-1 error:An error occurred (InvalidParameterValueException) when calling the CreateFunction operation: The role defined for the function cannot be assumed by Lambda.
2018-03-05 09:23:11,595: c7n_org:ERROR Exception running policy:ec2-tag-running account:account-302198017038 region:cn-north-1 error:An error occurred (InvalidParameterValueException) when calling the CreateFunction operation: The role defined for the function cannot be assumed by Lambda.
2018-03-05 09:23:11,606: c7n_org:INFO Policy resource counts Counter()

Q: where to define lambda excution role for B and C? There must be somewhere in policy file to define it. Or its not avaiable for lambda(cloudtrail) to be excuted in multi accounts?

**Comment by thisisshi** [comment]
It looks like the role you're assuming as doesn't have `lambda:CreateFunction` permission, I would check the IAM policy for that user or role

**Comment by kapilt** [comment]
per the error messgae, the role your using doesn't have a trust document allowing for use by lambda

**Comment by youzhoutai** [comment]
thank you two guys, actually i granted full lambda access for this role(CustodianEC2Tagging). i think there should be 2 types of roles for using c7n-org toll to implement lambda function. The first role for across account  authorization, such as lambda creation in this case, and the policy defined in configuration file as above step 1. but when we create lambda, one lambda excution role must be required, this is the second role. And this role should be difined in policy file above step 2, i am not sure if multi accounts could be defined in this policy file. maybe we can add some paras for lambda excution roles in the configuration file? i just take url below for reference.



http://capitalone.github.io/cloud-custodian/docs/policy/lambda.html

**Comment by kapilt** [comment]
custodian will default to use an assumed role for provisioning the lambda, the role specification in the lambda allows for specifying the two values separately, also there is variable interpolation on account id there to use a single policy with explicit role specification across multiple accounts.

**Comment by youzhoutai** [comment]
could u give one example about configuration file and policy file please? didnt get it. thank you.
